src/image.rs

```diff
@@ -921,6 +921,65 @@ pub trait GenericImageView {
         self.get_pixel(x, y)
     }
 
+    /// Returns the pixel located at (x, y) if the location is [`in_bounds`],
+    /// otherwise `None` is returned.
+    /// 
+    /// Location coordinates can be negative.
+    ///
+    /// # Safety
+    ///
+    /// The [`in_bounds`] check must work as expected.
+    /// 
+    /// [`in_bounds`]: #method.in_bounds
+    fn checked_get_pixel(&self, x: i64, y: i64) -> Option<Self::Pixel> {
+        if x.is_negative() || y.is_negative() {
+            None
+        } else {
+            match (u32::try_from(x), u32::try_from(y)) {
+                (Ok(x), Ok(y)) => {
+                    if self.in_bounds(x, y) {
+                        Some(unsafe { self.unsafe_get_pixel(x, y) })
+                    } else {
+                        None
+                    }
+                },
+                _ => None
+            }
+        }
+    }
+
+    /// Returns the pixel located at (x, y) if the location is [`in_bounds`],
+    /// otherwise the nearest bound pixel is returned.
+    /// 
+    /// Location coordinates can be negative.
+    /// 
+    /// [`in_bounds`]: #method.in_bounds
+    fn saturating_get_pixel(&self, x: i64, y: i64) -> Self::Pixel {
+        #[inline(always)]
+        fn convert(value: i64, bound: u32) -> u32 {
+            value.is_positive()
+                .then(|| {
+                    match u32::try_from(value) {
+                        Ok(value) => {
+                            if value < bound {
+                                value
+                            } else {
+                                bound - 1
+                            }
+                        },
+                        Err(_) => bound,
+                    }
+                }).unwrap_or(0u32)
+        }
+
+        unsafe {
+            self.unsafe_get_pixel(
+                convert(x, self.width()),
+                convert(y, self.height())
+            )
+        }
+    }
```

Comment:
> Also worth double checking that zero-sized images are either prohibited by GenericImage or properly handled by this patch

They are not. Panic in those cases would be permissible behavior but may it's not desirable. The methodical evaluation of how we wrote the trait is, however:

> [Safety](https://docs.rs/image/latest/image/trait.GenericImageView.html#safety)
> * The coordinates must be [in_bounds](https://docs.rs/image/latest/image/trait.GenericImageView.html#method.in_bounds) of the image.

Note that `in_bounds` is both safe, and the trait itself is safe. Thus, we can't make any assumption about its implementation other than the method signature. Since we haven't called the method here either, we can not argue to have fulfilled the precondition properly. I don't think this method possible in a generic, sound manner either way.

However, we _can_ add specialized implementations of these two new methods for `ImageBuffer`. Since the `in_bounds` method is local, we can use its implementation details and rely on them being correct.

Reply:
This is similar to `get_pixel_unchecked`. Note that the _generic_ implementation provided by the trait does not in fact call any unsafe method, it defaults to a purely safe fallback. This is correct and intended; since there's nothing `unsafe` about the trait, the implementor has no obligations to 'do the right thing'. There's no inherent property of `in_bounds` that the generic implementation could rely on for any effect; so calling the safe fallback is indeed the best useful implementation we can provide.

Note to self: `ImageBuffer` is still dubious if the user chooses a non-default container. Damn.